activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb

```diff
@@ -749,9 +749,34 @@ def configure_connection
           if @config[:timeout] && @config[:retries]
             raise ArgumentError, "Cannot specify both timeout and retries arguments"
           elsif @config[:timeout]
-            @raw_connection.busy_timeout(self.class.type_cast_config_to_integer(@config[:timeout]))
+            timeout = self.class.type_cast_config_to_integer(@config[:timeout])
+            raise TypeError, "Timeout must be an integer" unless timeout.is_a?(Integer)
+
+            timeout_seconds = timeout.fdiv(1000)
+            retry_interval = 6e-5 # 60 microseconds
+
+            # Custom busy_handler to release the GVL while wait for a connection.
+            raw_connection.busy_handler do |count|
+              timed_out = false
+              # keep track of elapsed time every 100 iterations (to lower load)
+              if (count % 100).zero?
+                # fail if we exceed the timeout value (captured from the timeout config option, converted to seconds)
+                timed_out = (count * retry_interval) > timeout_seconds
```

Comment:
Sounds really weird to me. You say it's to "minimize the ruby execution", but in the end with the extra modulo etc, I'm pretty sure you are doing as much work as if you'd just precomputed the `max_count` and compared it.

But even without precomputing, this amount of code is absolutely trivial.

Reply:
My understanding of what you mean by "precomputing the `max_count` and comparing it" would be the `retries` mechanism recently added. And this is precisely the reason I added this option. It has the benefits of doing the bare minimum amount of work in Ruby land, but doing work in Ruby land allows for the GVL to be released across retrying connections.

After getting some community feedback though, I realized that it does have the major downside of providing no real intuition for what number to put as the `max_count`/`retries`. One [example](https://github.com/rails/rails/pull/49352#issuecomment-1776227548) required 1M+ retries to get stable performance.

I think the semantics and intuition of `timeout` are vastly superior to `retries`. I also think that having some cap so that completely broken connections do eventually error is necessary. But, I also want "normal" connections, in "normal" multi-threaded contexts to not error at all. So, the goal here is to find the middle ground that achieves these goals. Whatever ways we can find to improve this `busy_handler` callback while still achieving these goals I am all in for.